engines/immortal/disk.h

```diff
@@ -0,0 +1,248 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef IMMORTAL_DISK_H
+#define IMMORTAL_DISK_H
+
+#include "common/memstream.h"
+#include "common/file.h"
+#include "common/debug.h"
+#include "common/error.h"
+
+/* Quick note about ProDOS:
+ * This disk code handles inactive, seedling, sapling, tree, and subdirectory files.
+ * It does *not* handle sparse files at the moment. If a sparse file exists, it may not
+ * be read correctly. It also does not do anything with Pascal files, but those should not
+ * matter for game engines anyway.
+ */
+
+namespace Immortal {
+
+// These values define for ProDOS how to read the file entry, and also whether it's a keyblock (if it is a directory header, it's the keyblock of that directory)
+enum FileType : char {
+	kFileTypeDead    = 0,
+	kFileTypeSeed    = 1,
+	kFileTypeSapling = 2,
+	kFileTypeTree    = 3,
+	kFileTypePascal  = 4,
+	kFileTypeSubDir  = 0x0D,
+	kFileTypeSubHead = 0x0E,
+	kFileTypeVolHead = 0x0F
+};
+
+/* File extensions for all the ProDOS supported file types
+ * NOTE: ProDOS user defined files are F1-F8. If they are ever required,
+ * they can be added to this enum.
+ */
+enum FileExt {
+	kFileExtNull     = 0,
+	kFileExtBad      = 1,
+	kFileExtTxt      = 4,
+	kFileExtBin      = 6,
+	kFileExtGfx      = 8,
+	kFileExtDir      = 0xF,
+	kFileExtDB       = 0x19,
+	kFileExtWord     = 0x1A,
+	kFileExtSpread   = 0x1B,
+	kFileExtSTART    = 0xB3,
+	kFileExtPascal   = 0xEF,
+	kFileExtPDCI     = 0xF0,
+	kFileExtPDRes    = 0xF9,
+	kFileExtIBProg   = 0xFA,
+	kFileExtIBVar    = 0xFB,
+	kFileExtAPSProg  = 0xFC,
+	kFileExtAPSVar   = 0xFD,
+	kFileExtEDASM    = 0xFE,
+	kFileExtSYS      = 0xFF
+};
+
+/* A ProDOS file simply contains meta data about the file and the ability to
+ * find and put together the data blocks that make up the file contents.
+ * This implements Common::ArchiveMember so that it can be used directly in
+ * the Archive methods in ProDOSDisk.
+ */
+
+class ProDOSFile : public Common::ArchiveMember {
+public:
+	ProDOSFile(char name[16], uint8 type, uint16 tBlk, uint32 eof, uint16 bPtr, Common::File *disk);
+	~ProDOSFile() {};											 // File does not need a destructor, because the file it reads from is a pointer to Disk, and Disk has a destructor
+
+	// -- These are the Common::ArchiveMember related functions --
+	Common::String getName() const override;                              // Returns _name
+	Common::SeekableReadStream *createReadStream() const override;        // This is what the archive needs to create a file
+	void getDataBlock(byte *memOffset, int offset, int size) const;       // Gets data up to the size of a single data block (512 bytes)
+	int parseIndexBlock(byte *memOffset, int blockNum, int cSize) const;  // Uses getDataBlock() on every pointer in the index file, adding them to byte * memory block
+
+	// For debugging purposes, just prints the metadata
+	void printInfo();
+
+private:
+	  char _name[16];
```

Comment:
Indentation is not correct here. If you want to align variable names, do it like this:

```C
char   _name[16];
uint8  _type;
uint16 _blocskPtr;
```